charmhub_lp_tools/charm_project.py

```diff
@@ -444,12 +445,45 @@ def get_revisions_for_bases(self,
                 if delete:
                     del revisions_[all_arch]
         # now just keep the highest revision for each arch.
-        highest_revisions = collections.defaultdict(int)
+        highest_revisions: Dict[str, int] = {}
         for k, v in revisions_.items():
             highest_revisions[k] = list(sorted(revisions_[k]))[-1]
-        # and collect the final set of revisions to keep.
-        revisions = set(v for v in highest_revisions.values())
-        return revisions
+        return highest_revisions
+
+    @staticmethod
+    def revisions_from_revisions_by_arch(
```

Comment:
these new methods need unit tests ;-)

Reply:
Technically they are tested by proxy as they are not mocked out. I'll explore the value of testing them separately.

Reply:
The main gain is to lock down the behavior for the future in case someone refactors the callers of revisions_from_revisions_by_arch() and this "testing by proxy" is lost.

Reply:
I've just had a look at almost no functions have tests, and those that do are tested poorly.  So there are a couple of options.  Add tests to this file properly; but that will seriously delay this patch going in, or add the tests as a series of separate patches/PRs that will gradually up the test coverage on the module.  Personally, I think this is important enough to merge quickly, as at the moment the function will clobber existing releases.

Reply:
Fair enough. We can tackle this aspect in issue #13 ;-)